framework/src/bcs/ADNodalBC.C

```diff
@@ -67,148 +68,92 @@ conversionHelper(const libMesh::VectorValue<ADReal> & value, const unsigned int
+      _undisplaced_assembly.processJacobianNoScaling(
+          conversionHelper(residual, i), dof_indices[i], _matrix_tags);
+}
 
-  mooseAssert(cached_rows.size() <= _set_components.size(),
-              "The number of dof indices must be less than the number of settable components");
+template <typename T>
+void
+ADNodalBCTempl<T>::computeResidual()
+{
+  const std::vector<dof_id_type> & dof_indices = _var.dofIndices();
+  if (dof_indices.empty())
+    return;
 
-  // Cache the user's computeQpJacobian() value for later use.
-  for (auto tag : _matrix_tags)
-    if (_sys.hasMatrix(tag))
-      for (std::size_t i = 0; i < cached_rows.size(); ++i)
-        if (_set_components[i])
-        {
-          _fe_problem.assembly(0, _sys.number())
-              .cacheJacobian(cached_rows[i],
-                             cached_rows[i],
-                             conversionHelper(residual, i).derivatives()[cached_rows[i]],
-                             tag);
-        }
+  const auto residual = computeQpResidual();
+
+  processResidual(residual, dof_indices);
 }
 
 template <typename T>
 void
-ADNodalBCTempl<T>::computeResidualAndJacobian()
+ADNodalBCTempl<T>::computeJacobian()
 {
+  const std::vector<dof_id_type> & dof_indices = _var.dofIndices();
+  if (dof_indices.empty())
+    return;
+
   const auto residual = computeQpResidual();
-  const std::vector<dof_id_type> & cached_rows = _var.dofIndices();
 
-  mooseAssert(cached_rows.size() <= _set_components.size(),
-              "The number of dof indices must be less than the number of settable components");
+  processJacobian(residual, dof_indices);
+}
 
-  // First the residuals
-  for (auto tag_id : _vector_tags)
-    if (_sys.hasVector(tag_id))
-      for (const auto i : index_range(cached_rows))
-        if (_set_components[i])
-          _sys.getVector(tag_id).set(cached_rows[i], raw_value(conversionHelper(residual, i)));
+template <typename T>
+void
+ADNodalBCTempl<T>::computeResidualAndJacobian()
+{
+  const std::vector<dof_id_type> & dof_indices = _var.dofIndices();
+  if (dof_indices.empty())
+    return;
 
-  // Cache the user's computeQpJacobian() value for later use.
-  for (auto tag_id : _matrix_tags)
-    if (_sys.hasMatrix(tag_id))
-      for (const auto i : index_range(cached_rows))
-        if (_set_components[i])
-          _fe_problem.assembly(0, _sys.number())
-              .cacheJacobian(cached_rows[i],
-                             cached_rows[i],
-                             conversionHelper(residual, i).derivatives()[cached_rows[i]],
-                             tag_id);
+  const auto residual = computeQpResidual();
+
+  processResidual(residual, dof_indices);
+  processJacobian(residual, dof_indices);
 }
 
 template <typename T>
 void
 ADNodalBCTempl<T>::computeOffDiagJacobian(const unsigned int jvar_num)
 {
+  // Only need to do this once because AD does all the derivatives at once
   if (jvar_num == _var.number())
     computeJacobian();
```

Comment:
are we worried it will double count then?
like a call to computeJacobian will have already happened for this variable?

Reply:
`NonlinearSystemBase::computeJacobianInternal` only makes calls to `computeOffDiagJacobian` for all coupling entries including ivar == jvar